infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -180,9 +187,36 @@ public void run() {
         } catch (Exception e) {
             log.error("Unhandled exception caught during log replication service discovery. Retry,", e);
         } finally {
-            if (runtime != null) {
-                runtime.shutdown();
-            }
+            runtime.cleanup(CorfuRuntime::shutdown);
```

Comment:
Is it the only place that clean up the runtime?

Reply:
Yes we cleanup runtime only once. The reason why it is wrapped in the SingletonResource is because we create the new lock clients for testing that also use this runtime.